pymc_experimental/distributions/timeseries.py

```diff
@@ -0,0 +1,198 @@
+    """
+
+    # TODO: Can this eval be avoided?
+    n, k = P.shape.eval()
+    if n != k:
+        raise ValueError(f'P must be square, found shape ({n}, {k})')
+
+    row_sums_all_one = at.allclose(P.sum(axis=1), 1.0)
+    if not row_sums_all_one.eval():
+        raise ValueError('All rows of P must sum to 1.')
+
+
+class DiscreteMarkovChainRV(SymbolicRandomVariable):
+    default_output = 1
+    _print_name = ('DiscreteMC', '\\operatorname{DiscreteMC}')
+
+    def update(self, node: Node):
+        # TODO: Do I need this?
+        return {node.inputs[-1]: node.outputs[0]}
+
+
+class DiscreteMarkovChain(Distribution):
+    r"""
+    A Discrete Markov Chain is a sequence of random variables
+    .. math::
+        \{x_t\}_{t=0}^T
+    Where transition probability P(x_t | x_{t-1}) depends only on the state of the system at x_{t-1}.
+
+    Parameters
+    ----------
+    P: tensor
+        Matrix of transition probabilities between states. Rows must sum to 1.
+        One of P or P_logits must be provided.
+    P_logit: tensor, Optional
+        Matrix of tranisiton logits. Converted to probabilities via Softmax activation.
+        One of P or P_logits must be provided.
+    steps: tensor
+        Length of the markov chain
+    x0: tensor or RandomVariable
+        Intial state of the system. If tensor, treated as deterministic.
+    """
+
+    rv_type = DiscreteMarkovChainRV
+
+    def __new__(cls, *args, steps, **kwargs):
+        # TODO: Allow steps to be None and infer chain length from shape?
+        # TODO: Dims breaks the RV
+
+        # Subtract 1 step to account for x0 given, better match user expectation of
+        # len(markov_chain) = steps
+        steps -= 1
+
+        steps = get_support_shape_1d(
+            support_shape=steps,
+            shape=None,
+            dims=kwargs.get('dims', None),
+            observed=kwargs.get('observed', None),
+            support_shape_offset=1
+        )
+
+        return super().__new__(cls, *args, steps=steps, **kwargs)
+
+    @classmethod
+    def dist(cls, P=None, logit_P=None, steps=None, x0=None, **kwargs):
+        steps = get_support_shape_1d(
+            support_shape=steps, shape=kwargs.get('shape', None), support_shape_offset=1
+        )
+        if steps is None:
+            raise ValueError("Must specify steps or shape parameter")
+        if P is None and logit_P is None:
+            raise ValueError('Must specify P or logit_P parameter')
+        if P is not None and logit_P is not None:
+            raise ValueError('Must specify only one of either P or logit_P parameter')
+
+        if logit_P is not None:
+            P = pm.math.softmax(logit_P, axis=1)
+        P = at.as_tensor_variable(P)
+        validate_transition_matrix(P)
+
+        # TODO: Can this eval be avoided?
+        n_states = P.shape[0].eval()
+
+        if not isinstance(x0, TensorVariable):
+            x0 = at.as_tensor_variable(x0).astype(intX)
+
+            # TODO: Can this eval be avoided?
```

Comment:
If the code wont just break loudly, you can wrap it in an Assert as well.

Reply:
I added a check for a dtype kwarg in the `dist` function and set it to `pytensor.config.floatX.replace('float', 'int')` if it's None. I was getting an error because the scan was returning `int32` variables, while the `init_dist` was returning int64. Is there a better way to handle this?

Reply:
I think you can set it to int64 as all our discrete distributions are of that type?

Reply:
Yeah i'm not sure what I thought the problem is. Removed all this and the code runs fine.